### DISCUSSION TITLE
[Installer] Is it possible to automate the entire download/installation process of the .net runtime?

### DISCUSSION THREAD

**Discussion by slxdy** [dotnet/runtime#108786] [Q&A] [answered]
Hey, I'm building a self-contained installer for an application that does require the .net runtime. Is there a(n official) way to automatically install the latest version of a specific major version? I have tried using the direct download link for the installers (from `download.visualstudio.microsoft.com`), but it refuses to send the content (the Content-Length is correct, but it won't send the entire content). Is there maybe an official way of doing this?

**Comment by slxdy** [comment]
I also got rate limited after only a few downloads, so that's fun.

**Comment by Falco20019** [comment, marked as answer]
Common URLs for latest versions:
* https://aka.ms/dotnet/6.0/dotnet-runtime-win-x64.exe (runtime)
* https://aka.ms/dotnet/6.0/dotnet-hosting-win.exe (ASP.NET Core runtime - Windows Hosting Bundle)
* https://aka.ms/dotnet/6.0/windowsdesktop-runtime-win-x64.exe (Windows Desktop runtime)
* https://aka.ms/dotnet/6.0/aspnetcore-runtime-win-x64.exe (ASP.NET Core runtime)
* https://aka.ms/dotnet/6.0/dotnet-sdk-win-x64.exe (SDK)

Preview builds: https://aka.ms/dotnet/6.0/preview/dotnet-sdk-win-x64.exe
Daily builds: https://aka.ms/dotnet/6.0/daily/dotnet-sdk-win-x64.exe
Specific builds: https://aka.ms/dotnet/6.0.100/dotnet-sdk-win-x64.exe

Of course you can also use any other major (like 7 or 8) in there. Not sure about .NET 9 though since it's in RC, but at least preview should work for it.

**Reply by Falco20019** [reply]
Did some tests. You can use `9.0/preview` to get the currently latest version. `9.0` is not live. Also specific build like `9.0.100-rc.2.24474.11` seems not to work.

**Comment by elgonzo** [comment]
I don't know what you tried and why it didn't work, but it's a simple HTTP download that doesn't require any trickery, no special treatment nor any special requests headers or cookies set.
Downloading the installer from `download.visualstudio.microsoft.com`, specifically
```
https://download.visualstudio.microsoft.com/download/pr/959afa1a-8afa-4d6f-982a-f09b9bb7392d/447a25613488738dbcf565e674db0669/dotnet-sdk-8.0.403-win-x86.exe
```
using `HttpClient.GetStream` or `HttpClient.GetStreamAsync` works just fine. Either the issue is with whatever tricky code you are employing or something in your general networking environment (spanning from your local network configuration to the operations of your ISP) interferes with your chosen approach...

**Reply by slxdy** [reply]
I used the link in the answer and it worked fine. It could be that I just got rate limited early, but either way, at least it's fixed

**Comment by EgorBo** [comment]
The simplest way is to use [dotnet-install](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) scripts, e.g.

Windows:
```ps1
Invoke-WebRequest https://dot.net/v1/dotnet-install.ps1 -OutFile dotnet-install.ps1 && .\dotnet-install.ps1 -Channel "8.0" -InstallDir LocalDotnet9
```

Linux:
```ps1
wget https://dot.net/v1/dotnet-install.sh && chmod +x dotnet-install.sh && ./dotnet-install.sh -Channel "8.0" -InstallDir LocalDotnet9
```
You can also use different parameters to get the latest/lts/whatever version you want
